### PR TITLE
DF simplify mapping

### PR DESCRIPTION
**Pull request by jbrezmorf**
- add method element_map, project_point
- complete unit test

### CODE REVIEW

File: src/fem/mapping_p1.cc [hunk 1]
```diff
@@ -276,6 +276,22 @@ void MappingP1<dim,spacedim>::fill_fe_side_values(const typename DOFHandlerBase:
     }
 }
 
+template<unsigned int dim, unsigned int spacedim>
+arma::vec MappingP1<dim,spacedim>::project_point(const arma::vec3 &point, const arma::mat &map) const
+{
+
+    if (dim == 0) return arma::ones(1);
```
Comment (jbrezmorf) [inline src/fem/mapping_p1.cc]: Used arma::Mat<double>::fixed<3,dim+1> and similarly for vectors.

File: src/fields/field_fe.cc [hunk 2]
```diff
@@ -239,8 +239,21 @@ void FieldFE<spacedim, Value>::interpolate(ElementDataCache<double>::ComponentDa
...
+			arma::vec projection;
+			switch (elm->dim()) {
+			case 1:
+				value_handler1_.point_projection(ele->centre(), projection, map, *elm);
+				break;
+			case 2:
+				value_handler2_.point_projection(ele->centre(), projection, map, *elm);
+				break;
+			case 3:
+				value_handler3_.point_projection(ele->centre(), projection, map, *elm);
```
Comment (jbrezmorf) [inline src/fields/field_fe.cc]: Method that decide if the point is in element or not. Move next if into the method and remove second test (projection close to centre). Even better move this decision method into Mapping.

File: src/io/observe.cc [hunk 3]
```diff
@@ -152,10 +156,11 @@ void ObservePoint::find_observe_point(Mesh &mesh) {
     for (unsigned int i_candidate=0; i_candidate<process_list.size(); ++i_candidate) {
         unsigned int i_elm=process_list[i_candidate];
         Element & elm = mesh.element[i_elm];
-        arma::mat map = elm.element_map();
+        arma::mat map;
+        arma::vec projection;
```
Comment (jbrezmorf) [inline src/io/observe.cc]: Make case according to dimension here and move rest of calculation on single element into auxiliary class template.
(insluding ObservePoint::snap method)

File: src/fem/mapping_p1.cc [hunk 4]
```diff
@@ -284,6 +284,38 @@ void MappingP1<0,3>::fill_fe_side_values(const DOFHandlerBase::CellIterator &cel
...
+
+
+template<unsigned int dim, unsigned int spacedim>
+arma::vec::fixed<dim+1> MappingP1<dim,spacedim>::clip_to_element(arma::vec::fixed<dim+1> &barycentric) {
+    return RefElement<dim>::clip(barycentric);
+}
+
+
+template<>
+arma::vec::fixed<1> MappingP1<0,3>::clip_to_element(arma::vec::fixed<1> &barycentric) {
```
Comment (jbrezmorf) [inline src/fem/mapping_p1.cc]: Try remove instance for MappingP1<0,3>.

File: src/io/observe.cc [hunk 5]
```diff
@@ -143,50 +230,44 @@ void ObservePoint::find_observe_point(Mesh &mesh) {
...
-
-            closed_elements.insert(i_elm);
-            // add all node neighbours to the next level list
-            for (unsigned int n=0; n < elm.n_nodes(); n++) {
-                for(unsigned int i_node_ele : mesh.node_elements[mesh.node_vector.index(elm.node[n])])
-                    candidate_list.push_back(i_node_ele);
+        // if element match region filter store it as observe element to the obs. point
+        if (elm.region().is_in_region_set(region_set)) {
+            // check that point is on the element
+            if ( point_projection(i_elm, projection_min, elm, ProjectionCases::update_if_in_elem) ) {
```
Comment (jbrezmorf) [inline src/io/observe.cc]: Join cases with and without clip_to_element.

**Comment by jbrezmorf** [comment]
Refactor search of observe point to use priority queue. Compute projection and TRUE distance for every ObservePointCandidate and put it into priority queue sorted by distance. 

Add unit test for point slightly out of a cube (should fail due to BIH tree checking of bounding boxes of candidate elements.)

File: src/io/observe.hh [hunk 6]
```diff
@@ -29,10 +55,14 @@ public:
...
             << "Failed to find the element containing the initial observe point.\n");
     TYPEDEF_ERR_INFO(EI_RegionName, std::string);
-    TYPEDEF_ERR_INFO(EI_NLevels, unsigned int);
     DECLARE_INPUT_EXCEPTION(ExcNoObserveElement,
             << "Failed to find the observe element with snap region: " << EI_RegionName::qval
-            << " close to the initial observe point. Using maximal number of neighbour levels: " << EI_NLevels::val << "\n");
+            << " close to the initial observe point. Change maximal distance of observe element." << "\n");
+
+    /// Helper enum specifies settings of update in point_projection method
+    enum ProjectionCases {
```
Comment (dflanderka) [inline src/io/observe.hh]: Check if enum is used, possibly remove.
Check the same way and remove unused local variables in find_observe_point method.

File: src/io/observe.cc [hunk 7]
```diff
@@ -74,60 +153,55 @@ ObservePoint::ObservePoint(Input::Record in_rec, unsigned int point_idx)
...
+void ObservePoint::update_projection(ObservePointData candidate_data)
 {
-    double dist = arma::norm(global_coords - input_point_, 2);
-    //cout << "dist: " << dist << endl;
-    if (dist < distance_) {
-        distance_ = dist;
-        element_idx_ = i_elm;
-        local_coords_ = local_coords;
-        global_coords_ = global_coords;
+    if (candidate_data.distance_ < observe_data_.distance_) {
```
Comment (jbrezmorf) [inline src/io/observe.cc]: Unecessary due to priority queue?

File: src/io/observe.cc [hunk 8]
```diff
@@ -141,110 +215,51 @@ void ObservePoint::find_observe_point(Mesh &mesh) {
     const BIHTree &bih_tree=mesh.get_bih_tree();
     vector<unsigned int> candidate_list, process_list;
     std::unordered_set<unsigned int> closed_elements(1023);
+    std::priority_queue< ObservePointData, std::vector<ObservePointData>, CompareByDist > candidate_queue;
 
     // search for the initial element
     bih_tree.find_point(input_point_, candidate_list);
```
Comment (jbrezmorf) [inline src/io/observe.cc]: Add parameter to return leaf elements without checking that point is contained in their bboxes

**Comment by dflanderka** [comment]
Done, branch can be merged to 3.0.0_dev

**Pull request merged**